Make Set methods obsolete

aaronfranke opened the pull request:
Mono: Declare all set methods obsolete and suggest using the constructors instead, fixes https://github.com/godotengine/godot/issues/30744

~~Core: Warn when using `set_axis_angle` and suggest using the constructor instead. Previously the constructor called the method, so I copied the logic.~~ We can delete the method in 4.0+. I looked into doing this for `set_euler` too, but the constructor doesn't work.

Partial follow-up to https://github.com/godotengine/godot/pull/21348

neikeq commented on modules/mono/glue/Managed/Files/Quat.cs:
```diff
@@ -95,6 +95,7 @@ public Quat Normalized()
             return this / Length;
         }
 
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("Set is deprecated. Use the Quat(" + nameof(real_t) + ", " + nameof(real_t) + ", " + nameof(real_t) + ", " + nameof(real_t) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Quat.cs:
```diff
@@ -103,16 +104,19 @@ public void Set(real_t x, real_t y, real_t z, real_t w)
             this.w = w;
         }
 
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("Set is deprecated. Use the Quat(" + nameof(Quat) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Quat.cs:
```diff
@@ -103,16 +104,19 @@ public void Set(real_t x, real_t y, real_t z, real_t w)
             this = q;
         }
 
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("SetAxisAngle is deprecated. Use the Quat(" + nameof(Vector3) + ", " + nameof(real_t) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Quat.cs:
```diff
@@ -103,16 +104,19 @@ public void Set(real_t x, real_t y, real_t z, real_t w)
             this = new Quat(axis, angle);
         }
 
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("SetEuler is deprecated. Use the Quat(" + nameof(Vector3) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Vector2.cs:
```diff
@@ -222,11 +222,13 @@ public Vector2 Round()
             return new Vector2(Mathf.Round(x), Mathf.Round(y));
         }
 
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("Set is deprecated. Use the Vector2(" + nameof(real_t) + ", " + nameof(real_t) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Vector2.cs:
```diff
@@ -222,11 +222,13 @@ public Vector2 Round()
             this.x = x;
             this.y = y;
         }
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("Set is deprecated. Use the Vector2(" + nameof(Vector2) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Vector3.cs:
```diff
@@ -248,12 +248,14 @@ public Vector3 Rotated(Vector3 axis, real_t phi)
             return new Basis(axis, phi).Xform(this);
         }
 
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("Set is deprecated. Use the Vector3(" + nameof(real_t) + ", " + nameof(real_t) + ", " + nameof(real_t) + ") constructor instead.", error: true)]
```

neikeq commented on modules/mono/glue/Managed/Files/Vector3.cs:
```diff
@@ -248,12 +248,14 @@ public Vector3 Rotated(Vector3 axis, real_t phi)
             this.y = y;
             this.z = z;
         }
+        [Obsolete("Deprecated, please use constructor and assignment instead.", true)]
```
```C#
[Obsolete("Set is deprecated. Use the Quat(" + nameof(Vector3) + ") constructor instead.", error: true)]
```

neikeq reviewed (changes requested):
I've commented how the Obsolete messages should be, to be consistent with the existing ones. I also think it's better to specify the exact constructor to use. The assignment part is not needed as people will understand it. Also, we must use named parameters for booleans.

I can't comment on the C++ API. I don't know if the changes are needed there.

akien-mga commented on core/math/quat.cpp:
```diff
@@ -232,6 +232,8 @@ Quat::operator String() const {
 }
 
 void Quat::set_axis_angle(const Vector3 &axis, const real_t &angle) {
+	ERR_EXPLAIN("set_axis_angle is deprecated. Use the Quat(Vector3 axis, float angle) constructor instead.");
```
You could remove the duplicated code below by a call to the constructor I guess.

akien-mga commented:
To be fair I'm not sure if it's really useful to deprecate this API in Quat. I can't comment about C#, but Quat has `set_euler()` and `set_axis_angle()`, which both make their mathematical implications clear. `Quat(Vector3)` and `Quat(Vector3, float)` on the other hand are a lot less obvious, and while the documentation can make the arguments expicit, once it's in your code, `quat = Quat(some_vec, some_float)` will be less explicit than `quat.set_axis_angle(some_vec, some_float)`.

aaronfranke commented:
@akien-mga Those methods in Basis aren't currently exposed (and also aren't in C#).

There are actual issues with these in C# that causes the code to silently fail, so if core is unchanged it would still be a good idea to remove them from the mono module.

neikeq reviewed (approved):
C# changes are good.

aaronfranke committed:
[Mono] Deprecate Set methods

These silently fail, so they should be removed. I accidentally added most of these last year, trying to make everything else consistent with Quat, sorry!

Also, a few tiny nitpicking changes are included, like whitespace and misspellings.

aaronfranke commented:
I removed the core Quat changes from this PR, so that it's just the C# changes and some nitpicking stuff. I have the core change saved elsewhere in case we want to revisit this later, if it's better to keep those methods in GDScript then that's fine even if it's inconsistent with C#.

akien-mga commented:
Thanks!

For the core changes, you can mention them in #16863.
